### PR TITLE
[YUNIKORN-1608] Configuration updates and storage for limit

### PR DESCRIPTION
**Pull request by manirajv06**
### What is this PR for?
Traverse the limits configuration (already passed through validation) for all queues recursively and set the user and groups limits in corresponding QueueTracker object.

### What type of PR is it?
* [ ] - Improvement

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-1608

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

**Comment by manirajv06** [comment]
Test failure is not related

**Comment by manirajv06** [comment]
Test failure is not related

**Review by wilfred-s — CHANGES_REQUESTED** [review]

**Comment by wilfred-s** [comment]
> Test failure is not related

I am also not sure that the test failures are not from this patch.
We have run other PR pre-commit flows in the last days and all tests pass.
This one failed on re-run also.

**Comment by manirajv06** [comment]
> > Test failure is not related
> 
> I am also not sure that the test failures are not from this patch. We have run other PR pre-commit flows in the last days and all tests pass. This one failed on re-run also.

Tests are passing.

**Review by wilfred-s — CHANGES_REQUESTED** [review]
Use a single line construct everywhere when checking errors for return:
```
if err = gt.setMaxApplications(0, queuePath); err != nil {
   return err
}
```

File: pkg/scheduler/ugm/queue_tracker.go [hunk 1]
```diff
@@ -115,6 +129,74 @@ func (qt *QueueTracker) decreaseTrackedResource(queuePath string, applicationID
...
+			zap.Int("total running applications", len(childQueueTracker.runningApplications)),
+			zap.Uint64("config max applications", count))
+		return fmt.Errorf("current running applications is greater than config max applications for %s", queuePath)
+	} else {
+		childQueueTracker.maxRunningApps = count
+	}
+	return nil
+}
+
+func (qt *QueueTracker) setMaxResources(resource *resources.Resource, queuePath string) error {
```
Comment (pbacsko) [inline pkg/scheduler/ugm/queue_tracker.go:166]: This method looks almost the same as `setMaxApplications()`, there are a lot of common things, could you extract it? Plus, there's missing coverage here, please address that.

File: pkg/scheduler/utilities_test.go [hunk 2]
```diff
@@ -376,9 +587,77 @@ func getTestUserGroup() security.UserGroup {
...
+	expectedQueuesMaxLimits["root"] = make(map[string]interface{})
+	expectedQueuesMaxLimits["root.default"] = make(map[string]interface{})
+	expectedQueuesMaxLimits["root"][maxresources] = resources.NewResourceFromMap(map[string]resources.Quantity{"memory": 10, "vcores": 10})
+	expectedQueuesMaxLimits["root.default"][maxresources] = resources.NewResourceFromMap(map[string]resources.Quantity{"memory": 5, "vcores": 5})
+	expectedQueuesMaxLimits["root"][maxapplications] = uint64(2)
+	expectedQueuesMaxLimits["root.default"][maxapplications] = uint64(1)
+	internalAssertUserGroupResource(t, userGroup, expected, expectedQueuesMaxLimits)
+}
+
+func internalAssertUserGroupResource(t *testing.T, userGroup security.UserGroup, expected *resources.Resource, expectedQueuesMaxLimits map[string]map[string]interface{}) {
```
Comment (pbacsko) [inline pkg/scheduler/utilities_test.go:600]: I'd rename this to `assertUserGroupResourceMaxLimits()`, or maybe even just `assertResourcesMaxLimits()` (and the existing can be changed to `assertResources()`). Otherwise I have to scan with my eyes to see why it's called "internal".

File: pkg/scheduler/utilities_test.go [hunk 3]
```diff
@@ -376,9 +587,77 @@ func getTestUserGroup() security.UserGroup {
...
+			if qApps, ok := gMaxApps[q]; ok {
+				assert.Equal(t, qApps, qMaxLimits[maxapplications].(uint64))
+			}
+		}
+	}
 	assert.Equal(t, resources.Equals(userResource, expected), true)
 	assert.Equal(t, resources.Equals(groupResource, expected), true)
 }
+
+func internalGetMaxResource(usage *dao.ResourceUsageDAOInfo, maxResources map[string]*resources.Resource) map[string]*resources.Resource {
```
Comment (pbacsko) [inline pkg/scheduler/utilities_test.go:645]: Nit: just call it `getMaxResource()`, it doesn't collide with an existing function.

File: pkg/scheduler/objects/queue.go [hunk 4]
```diff
@@ -135,7 +135,6 @@ func NewConfiguredQueue(conf configs.QueueConfig, parent *Queue) (*Queue, error)
 		sq.mergeProperties(parent.getProperties(), conf.Properties)
 	}
 	sq.UpdateQueueProperties()
-
```
Comment (pbacsko) [inline pkg/scheduler/objects/queue.go:138]: Nit: unnecessary change here.

File: pkg/scheduler/ugm/manager.go [hunk 5]
```diff
@@ -73,11 +80,66 @@ func (m *Manager) IncreaseTrackedResource(queuePath string, applicationID string
...
+			if err != nil {
+				log.Logger().Warn("Problem in setting the limit max applications settings.",
+					zap.String("user", user.User),
+					zap.String("queue path", configQueuePath),
+					zap.Uint64("limit max applications", maxApps),
+					zap.Error(err))
+				return fmt.Errorf("unable to set the max applications. user: %s, queuepath : %s, applicationid: %s, usage: %s, reason: %w",
+					user.User, configQueuePath, applicationID, usage.String(), err)
+			}
+			maxResources, maxResourcesErr := config[maxresources].(map[string]string)
```
Comment (pbacsko) [inline pkg/scheduler/ugm/manager.go:113]: Nit: just simply `ok` instead of  `maxResourcesErr`, that's the idiomatic Go approach.

File: pkg/scheduler/ugm/manager.go [hunk 6]
```diff
@@ -73,11 +80,66 @@ func (m *Manager) IncreaseTrackedResource(queuePath string, applicationID string
...
+			if !maxResourcesErr {
+				log.Logger().Warn("Problem in setting the limit max resources settings. Unable to cast the value from interface to resource",
+					zap.String("user", user.User),
+					zap.String("queue path", configQueuePath),
+					zap.Any("limit max resources", maxResources))
+				return fmt.Errorf("unable to set the max resources. user: %s, queuepath : %s, applicationid: %s",
+					user.User, configQueuePath, applicationID)
+			}
+			resource, resourceErr := resources.NewResourceFromConf(maxResources)
```
Comment (pbacsko) [inline pkg/scheduler/ugm/manager.go:122]: Nit: just simply `ok` instead of  `maxResourcesErr`, that's the idiomatic Go approach.

Comment (pbacsko) [inline pkg/scheduler/ugm/manager.go:122]: There are some codecov complaints here, please address those if relevant (I'm generally in favor of checking the negative paths).

File: pkg/scheduler/ugm/manager.go [hunk 7]
```diff
@@ -314,15 +432,278 @@ func (m *Manager) isGroupRemovable(gt *GroupTracker) bool {
...
+	m.Lock()
+	defer m.Unlock()
+
+	// clear the local limit config maps before processing the limit config
+	m.userLimitsConfig = make(map[string]map[string]map[string]interface{})
+	m.groupLimitsConfig = make(map[string]map[string]map[string]interface{})
+	return m.internalProcessConfig(config, queuePath)
+}
+
+func (m *Manager) internalProcessConfig(cur configs.QueueConfig, queuePath string) error {
```
Comment (pbacsko) [inline pkg/scheduler/ugm/manager.go:445]: This method is excessively long. Can't you split it up to reasonable parts & name them properly?

**Comment by manirajv06** [comment]
@wilfred-s Can you also take a final look?

**Review by wilfred-s — APPROVED** [review]
LGTM

**Comment by manirajv06** [comment]
Filed https://issues.apache.org/jira/browse/YUNIKORN-1815 based on discussion.
